eclair-core/src/main/scala/fr/acinq/eclair/router/BalanceEstimate.scala

```diff
@@ -0,0 +1,263 @@
+   *
+   * @param amount                the amount that we knew we could send or not send at time t
+   * @param successProbabilityAtT probability that we could relay amount at time t (usually 0 or 1)
+   * @param t                     time at which we knew if we could or couldn't send amount
+   * @return                      the probability that we can send amount now
+   */
+  private def decay(amount: MilliSatoshi, successProbabilityAtT: Double, t: TimestampSecond): Double = {
+    val decayRatio = 1 / math.pow(2, (TimestampSecond.now() - t) / halfLife)
+    val baseline = 1 - amount.toLong.toDouble / toMilliSatoshi(totalCapacity).toLong.toDouble
+    baseline * (1 - decayRatio) + successProbabilityAtT * decayRatio
+  }
+
+  def otherSide: BalanceEstimate = BalanceEstimate(totalCapacity - high, highTimestamp, totalCapacity - low, lowTimestamp, totalCapacity, halfLife)
+
+  def couldNotSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    if (amount < high) {
+      if (amount > low) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        // the balance is actually below `low`, we discard our lower bound
+        copy(low = MilliSatoshi(0), lowTimestamp = timestamp, high = amount, highTimestamp = timestamp)
+      }
+    } else {
+      /* We already expected not to be able to relay that amount as it above our upper bound. However if the upper bound
+       * was old enough that replacing it with the current amount decreases the success probability for `high`, then we
+       * replace it.
+       */
+      val pLow = decay(low, 1, lowTimestamp)
+      val pHigh = decay(high, 0, highTimestamp)
+      val x = low + (high - low) * (pLow / (pLow - pHigh))
+      if (amount <= x) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        copy()
+      }
+    }
+
+  def couldSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    otherSide.couldNotSend(totalCapacity - amount, timestamp).otherSide
+
+  def didSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate = {
+    val newLow = (low - amount) max MilliSatoshi(0)
+    val newHigh = (high - amount) max MilliSatoshi(0)
+    /* We could shift everything left by amount without changing the timestamps (a) but we may get more information by
+     * ignoring the old high (b) if if has decayed too much. We try both and choose the one that gives the lowest
+     * probability for high.
+     */
+    val a = copy(low = newLow, high = newHigh)
+    val b = copy(low = newLow, high = (totalCapacity - amount) max MilliSatoshi(0), highTimestamp = timestamp)
+    if (a.canSend(newHigh) < b.canSend(newHigh)) {
+      a
+    } else {
+      b
+    }
+  }
+
+  def addChannel(capacity: Satoshi): BalanceEstimate = copy(high = high + toMilliSatoshi(capacity), totalCapacity = totalCapacity + capacity)
```

Comment:
I'm still making the same comment I initially made on the first iteration of this feature: I think it really doesn't make sense to add channel capacities, I believe we should take the `max` of the channel capacities instead (which means we need to keep a list of channel capacities to correctly handle removing channels).

If nodes have more than a few channels between them, the `high` bound will be completely unreasonable and the decay will tend to restore it way too quickly to an impossible amount (since the balance estimates will be used in the context of a single HTLC, not for the whole flow of a payment).

Also, you're not even using it consistently: `BalanceEstimates.addChannel` doesn't add the capacities whereas other functions do...which shows we probably need more unit tests :wink: 

What about having the balance estimate contain a `capacities: Seq[MilliSatoshi]`? This also lets us know how many channels exist between these peers, which may help us choose to use this pair of nodes for more than one HTLC when using MPP.